### ISSUE TITLE
style: Write a static code analyzer for sclang

### ISSUE THREAD

**Issue by snappizz** [supercollider/supercollider#2915]
## Issue

we have been having a few arguments about enforcing style. a lot of small style decisions can be enforced by a static code analyzer.

this is a technical proposal, not an organizational one. if and when we have a style enforcer, there's still plenty of room to decide how strictly we want to foist it upon contributors.

## Description

some uncontroversial style rules i suggest are:

- outright syntax error detection
- enforce proper use of whitespace: no trailing spaces, spaces around binary operators, no spaces in parens, spaces after but not before commas, etc.
- disallow use of global variables `a-z` in the class library
- disallow use of `~variables` directly inside methods (ok if inside function)
- disallow calls to private methods except on `this`
- disallow abuse of assignment operator as an expression, e.g. `foo.value(bar = "baz")`
- disallow confusing order of math operations, e.g. `3 + 4 * 5`
- disallow error-prone chains of methods on top of partial application: `_.squared.sqrt`
- disallow use of `_` with keyword arguments
- disallow unrecognized backslash escaping of characters in strings that could be a regex-related mistake, e.g. `"example\.com".matchRegexp("example~com")`
- require a statement such as `this.primitiveFailed` after a primitive
- forgetting commas between arguments in `|argument lists|`

i like [jslint](http://jslint.com/)'s approach to dangerous constructs that are sometimes needed anyway: you have to place a `/*jslint */` directive at the top of the file to explicitly allow that feature. this forces you (and anyone reviewing your code) to stop and consider your decision before you disable the warning. i suggest this for:

- calls to the `interpret` method
- use of `~variables`, even if inside functions

i would prefer if the comments on this issue not devolve into bikeshedding over the style guide, so please only use it for uncontroversial proposals.

## Implementation

i could definitely take this on once i have the time, and i could write it in sclang. or maybe another language, idk.

i would much rather write a new lexer and parser rather than piggyback off our existing ones. writing a new one gives us the ability to cross-check things with the existing language. i have written parsers and lexers before, and believe me, you learn so much about the edge cases in the syntax you're parsing. i cannot wait to discover new tokenization horrors in `yylex`.

furthermore, an sclang parser is useful for many things other than style enforcement, so i think the general SC user base could benefit from this.

**Comment by brianlheim** [comment]
What's the expected product and use case here? It actually sounds like two projects to me, static analysis and linting. We should probably separate those concepts right away, because they require different levels of complexity (plaintext analysis vs AST analysis). IMO, this would be better as a separate project outside of the main repo, but could perhaps be brought in as a submodule.

**Comment by snappizz** [comment]
you need to make an AST in order to lint, or at least how i'm imagining the linter.

**Comment by brianlheim** [comment]
yeah, ok, I can see that. thoughts re: doing it elsewhere?

**Comment by snappizz** [comment]
i'm fine with a quark, but i feel that this is both realistically attainable and relevant to the supercollider project to the point that this issue can stay open. what do you think?

**Comment by brianlheim** [comment]
Sure! So, is the idea to keep it open until it's implemented in at least one spot? I'm totally fine with that.

**Comment by telephon** [comment]
Fine with most, thanks for assembling them.

>disallow confusing order of math operations, e.g. 3 + 4 * 5

As such, that might be overly strict. The reason behind having no operator precedence in sclang is that the rules would be really complicated (example: `3 + 4 ** 5` or `3 + 4 gcd: 8`). But we can suggest that parentheses can be used for clarity?

Your two extra suggestions (interpret and dynamic scope) shouldn't by any means be understood as indicating "bad style". They are not, they have utter dignity.

You are right that sometimes they are, but I am not sure if I could make it *explicit* when exactly.

**Comment by snappizz** [comment]
@telephon yes, my proposal is that parentheses are required in the case of `(a <b> c) <d> e` if standard mathematical notation has `<b>` at a lower precedence than `<d>`.

i don't just want code to look pretty, i want to prevent potential errors. the point of `interpret` and dynamic scope in the class library is that, although use of those features isn't automatically bad, they are common beginner mistakes and could be a sign of bad code.

the SCA would require the user to put something like `/*sclint interpret */` at the top of the file for it to pass. this serves two functions: first, it gets the author to stop and consider their decision to use these features. second, it gives signals to future readers of the code: they might see that flag and think "huh, why the heck would someone need to use `interpret` in `Color`?!"

**Comment by jamshark70** [comment]
> enforce proper use of whitespace: no trailing spaces, spaces around binary operators, no spaces in parens, spaces after but not before commas, etc.

Seems simple enough on its face, but there's likely to be disagreement over details, no?

- `[0, 1, 2]` or `[ 0, 1, 2 ]`?
- `[[0, 1], 2]` or `[ [ 0, 1 ], 2 ]`?
- `{ SinOsc.ar }.play` or `{SinOsc.ar}.play`?
- `if(condition)` or `if (condition)`?
- `10.do { |i| ... }` or `10.do { | i | ... }`?

... probably some others I can't think of right now. I prefer the former for each of these, but I've seen code in the wild with all of the latter choices.

In particular, the last one, I *very very strongly dislike* spaces around the pipes, but #2913 was opened initially with an almost casual assumption that the spaces are desirable. I justify my preference by noting that we prefer spaces around binary operators (and I agree with that), but if we encourage spaces around arg-pipes, there is no immediate visual distinction between pipe = arglist or pipe = bit-or. IMO that would be a mistake. `{ |a = 1| a }` -- obvious that `|` is a delimiter. `{ | a = 1 | a }` -- when I see `1 | a` here, I have to look backward and see if there's a matching/opening `|` -- it takes more work to read and understand the code.

I'd also suggest an exception to the binary operator rule: `array[i-1]` does not offend my eye, although `x[index-1]` would (because of the width of `i` vs `index`).

> disallow abuse of assignment operator as an expression

It's very slightly more efficient in some cases, though. It's not that much more efficient to warrant dropping the rule (and as noted, users can opt for exceptions per-file), but:

```
Array.fill(15, { bench { var x = 0; while { x = x + 1; x <= 10000000 } } }).median;
-> 0.31879949099999

Array.fill(15, { bench { var x = 0; while { (x = x + 1) <= 10000000 } } }).median;
-> 0.29904191399999

0.31879949099999 / 0.29904191399999
-> 1.0660695911677
```

'bout 6%.

**Comment by snappizz** [comment]
we don't need to enforce a detail if we haven't reached consensus on it. it's most important to catch style issues that are prone to error, like `_.pow(2).neg`.

**Comment by brianlheim** [comment]
> but #2913 was opened initially with an almost casual assumption that the spaces are desirable.

Sorry! I opened it thinking we would only be discussing arg vs ||. I didn't want to open too many points of style at once. But, I am happy that the conversation moved beyond that!

**Comment by brianlheim** [comment]
More and more I think this is becoming necessary. I would like to devote some time to writing this, but I definitely will not be able to get to it until this summer.

**Comment by scztt** [comment]
When we finish a compile, we have an AST that's fairly navigable. If someone wanted to, it would be a pretty straightforward task to serialize the 15 or so AST nodes we have, such that they could be deserialized by either sclang (for linting and introspection) or an IDE later on.

We could skip the serialization part even, if we rebuild our AST nodes such that they are sclang-compatible objects (e.g. convert members to PyrSlot's) - then, they could simply hang around and be used by sclang after launch.

Building a lexer/parser in a different language and maintaining it in parallel to the existing one seems like the biggest and heaviest kind of technical debt - are there benefits that outweigh the cost of carrying two complex high-opacity components into the future instead of just one?

**Comment by brianlheim** [comment]
> Building a lexer/parser in a different language and maintaining it in parallel to the existing one seems like the biggest and heaviest kind of technical debt - are there benefits that outweigh the cost of carrying two complex high-opacity components into the future instead of just one?

Maybe I'm missing something – how would you get anything out of the AST that you could feed to a linter, without writing at least most of a new parser? The AST is fine for most of the problems already listed, but not for whitespace and formatting rules AFAICT. Maybe we should break out a separate issue for a linter so that this can just be about static analysis.

**Comment by scztt** [comment]
Having tried to rewrite the busted IDE indentation rules multiple times without success, there are a handful of cases (in sclang at least) that require knowledge about the AST to solve, over and above brute-force tokenization. And these cases aside, my more successful attempts at correct indentation end up looking a lot like an AST - which is a clue that I'm probably creating a parallel data structure of information we already have elsewhere.
The thing to consider: if I change sclang's `lang11d` file and corresponding AST, is it going to require changes in the linter / formatting code? If so, they probably share a single common data structure.

**Comment by scztt** [comment]
An example of what I mean - getting the following to indent correctly when split to multiple lines basically requires building a nested structure containing with each nested {}, [], ||, () expression, as well as their start and end char position. That's already in the AST - and the code to extract the right info from the source basically amounts to a duplicated parser with some details removed.
`f = {
	|a = ({|b = ([{ |c| c.postln }, {[b]} ])| b.postln })|
}`

**Comment by brianlheim** [comment]
> there are a handful of cases (in sclang at least) that require knowledge about the AST to solve, over and above brute-force tokenization. And these cases aside, my more successful attempts at correct indentation end up looking a lot like an AST - which is a clue that I'm probably creating a parallel data structure of information we already have elsewhere.

OK! If/when I start working on this I'll definitely spend some time thinking about how I can avoid duplication. I'm glad all this stuff is coming up early on, before implementation.

> That's already in the AST - and the code to extract the right info from the source basically amounts to a duplicated parser with some details removed.

That's a good point. I don't think start and end positions are maintained in the AST - at least, when I was working in some of that code yesterday I didn't see any. But I think with some slight modification that could be added.

Something that isn't immediately clear to me, however, is whether you could still get an AST when the file produces a parsing error. Ideally, you can always hit "indent" and get a sensible answer. If we went with the approach of pulling info from the AST, would that still be possible?

**Comment by jamshark70** [comment]
> Having tried to rewrite the busted IDE indentation rules multiple times without success, there are a handful of cases (in sclang at least) that require knowledge about the AST to solve, over and above brute-force tokenization.

Another use case: the IDE can use the tree for a smart navigation mode, where you can move through the tree by arrow keys and the text selection follows. My live coding system can do this in a TextView and it's really helpful. (Ctrl-arrows usually follow word semantics, which is often silly for programming languages, and almost totally useless for my punctuation-heavy live coding syntax.) With that, you could left/right-arrow through array *elements*, no matter their length, because they are all same-level siblings in the tree.

Orthogonal to this issue but I'm just mentioning briefly, syntax-tree selection navigation is *really cool*.

**Comment by telephon** [comment]
would it be possible to trace back the position in the AST that has created a given object?

**Comment by snappizz** [comment]
no resolution.
